### DISCUSSION TITLE
아래 코드처럼 외부 함수 자체가 중첩된 함수인 경우, 결과가 어떻게 되는 지 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#86] [클로저]
- 아래 코드처럼 외부 함수 자체가 중첩된 함수인 경우, 결과가 어떻게 되는 지 설명해주세요.
  ```js
  const e = 10;
  function sum(a) {
    return function (b) {
      return function (c) {
        return function (d) {
          return a + b + c + d + e;
        };
      };
    };
  }
  
  console.log(sum(1)(2)(3)(4));
  
  ```
  - 이 함수들이 전부 외부 함수의 범위에 접근할 수 있는 이유는 무엇인가요?

**Comment by publdaze** [comment]
> - 아래 코드처럼 외부 함수 자체가 중첩된 함수인 경우, 결과가 어떻게 되는 지 설명해주세요.
> 
>   ```js
>   const e = 10;
>   function sum(a) {
>     return function (b) {
>       return function (c) {
>         return function (d) {
>           return a + b + c + d + e;
>         };
>       };
>     };
>   }
> 
>   console.log(sum(1)(2)(3)(4));
>   ```
> - 이 함수들이 전부 외부 함수의 범위에 접근할 수 있는 이유는 무엇인가요?
- 각 중첩함수는 렉시컬 스코프에 따라 함수가 정의되는 위치에 따라 스코프가 결정됨
- 정의되는 위치 기반으로 식별자를 검색하기 위해 렉시컬 환경을 탐색하게 될 텐데
- 가장 하위 중첩 함수의 렉시컬 환경에 존재하지 않는 식별자는 스코프 체인을 따라 외부 렉시컬 환경을 타고 올라가며 검색 가능
- 각 외부 함수는 클로저에 의해 생명주기가 끝나도 렉시컬 환경이 제거되지 않고 남아있어서 식별자 검색이 가능 함
- 최종적으로 20이 출력됨

**Comment by pipisebastian** [comment]
> * 아래 코드처럼 외부 함수 자체가 중첩된 함수인 경우, 결과가 어떻게 되는 지 설명해주세요.
>   ```js
>   const e = 10;
>   function sum(a) {
>     return function (b) {
>       return function (c) {
>         return function (d) {
>           return a + b + c + d + e;
>         };
>       };
>     };
>   }
>   
>   console.log(sum(1)(2)(3)(4));
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   
>   * 이 함수들이 전부 외부 함수의 범위에 접근할 수 있는 이유는 무엇인가요?


렉시컬 스코프와 스코프 체인 때문입니다.
렉시컬 스코프란, 함수가 정의된 위치에 따라 상위스코프가 결정되는 방식입니다. 각 내부함수의 상위스코프는 해당 내부함수를 감싸는 외부함수인데, 각 외부함수들은, 클로저에 의해 렉시컬환경을 계속 유지합니다. 그렇기 때문에, `return a + b + c + d + e;`에서 a,b,c,d 는 클로저에 의해 참조가 가능하고, 각각은 스코프체인을 통해 상위 스코프로 올라가서 식별자를 찾습니다. 또한, e는 전역변수이므로 스코프 체인을 통해 참조가 가능합니다.
